Issues with GCC 12.1

Hi

I need the following patch to compile with GCC 12.1

```
From 752fc3bd9a971858dbb7b51a0ef0289232e67e58 Mon Sep 17 00:00:00 2001
From: Mike Lothian <mike@fireburn.co.uk>
Date: Sat, 14 May 2022 02:01:18 +0100
Subject: [PATCH] Add functional include

---
tls/include/adb/tls/tls_connection.h | 1 +
1 file changed, 1 insertion(+)

diff --git a/tls/include/adb/tls/tls_connection.h
b/tls/include/adb/tls/tls_connection.h
index bc5b98ab..a112756c 100644
--- a/tls/include/adb/tls/tls_connection.h
+++ b/tls/include/adb/tls/tls_connection.h
@@ -19,6 +19,7 @@
#include <stddef.h>
#include <stdint.h>

+#include <functional>
#include <string_view>
#include <vector>
```

It also warns about:

```
 * QA Notice: Package triggers severe warnings which indicate that it
 *            may exhibit random runtime failures.
 * /var/tmp/portage/dev-util/android-tools-31.0.3_p1/work/android-tools-31.0.3p1/vendor/selinux/libselinux/src/label_file.h:289:25: warning: ‘free’ called on pointer ‘mmap_area_89->next_addr’ with nonzero offset 4 [-Wfree-nonheap-object]
```
Which is fixed with https://github.com/SELinuxProject/selinux/commit/6e5d16a012e1845774da106846b87090d39ea8f3

--- a/patches/adb/0021-tls-include-functional.patch
+++ b/patches/adb/0021-tls-include-functional.patch
@@ -1,0 +1,22 @@
+From 9d0394f52324e7ccc62c67ce34cfe6ef9e85f043 Mon Sep 17 00:00:00 2001
+From: =?UTF-8?q?S=C3=B6ren=20Tempel?= <soeren+git@soeren-tempel.net>
+Date: Wed, 18 May 2022 01:02:57 +0200
+Subject: [PATCH] tls: include <functional>
+
+Fixes compilation with GCC 12.1.
+---
+ tls/include/adb/tls/tls_connection.h | 1 +
+ 1 file changed, 1 insertion(+)
+
+diff --git a/tls/include/adb/tls/tls_connection.h b/tls/include/adb/tls/tls_connection.h
+index bc5b98ab..a112756c 100644
+--- a/tls/include/adb/tls/tls_connection.h
++++ b/tls/include/adb/tls/tls_connection.h
+@@ -19,6 +19,7 @@
+ #include <stddef.h>
+ #include <stdint.h>
+ 
++#include <functional>
+ #include <string_view>
+ #include <vector>
+ 

--- a/patches/boringssl/0012-CMakeLists.txt-Disable-Werror-by-default.patch
+++ b/patches/boringssl/0012-CMakeLists.txt-Disable-Werror-by-default.patch
@@ -1,0 +1,24 @@
+From 9ecab80323875e95b08b5e6f327e4d5bd5d70478 Mon Sep 17 00:00:00 2001
+From: =?UTF-8?q?S=C3=B6ren=20Tempel?= <soeren+git@soeren-tempel.net>
+Date: Sat, 21 May 2022 11:47:32 +0200
+Subject: [PATCH] CMakeLists.txt: Disable -Werror by default
+
+Boringssl regularly fails to compile with newer GCC versions due
+to new warnings introduced by GCC upstream.
+---
+ CMakeLists.txt | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index f58e853cd..b1fec9616 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -117,7 +117,7 @@ endif()
+ if(CMAKE_COMPILER_IS_GNUCXX OR CLANG)
+   # Note clang-cl is odd and sets both CLANG and MSVC. We base our configuration
+   # primarily on our normal Clang one.
+-  set(C_CXX_FLAGS "-Werror -Wformat=2 -Wsign-compare -Wmissing-field-initializers -Wwrite-strings -Wvla")
++  set(C_CXX_FLAGS "-Wformat=2 -Wsign-compare -Wmissing-field-initializers -Wwrite-strings -Wvla")
+   if(MSVC)
+     # clang-cl sets different default warnings than clang. It also treats -Wall
+     # as -Weverything, to match MSVC. Instead -W3 is the alias for -Wall.

--- a/patches/selinux/0002-libselinux-store_stem-do-not-free-possible-non-heap-.patch
+++ b/patches/selinux/0002-libselinux-store_stem-do-not-free-possible-non-heap-.patch
@@ -1,0 +1,59 @@
+From 0797bc2697e8ae9d2e3527bf0521274b6ff94473 Mon Sep 17 00:00:00 2001
+From: =?UTF-8?q?Christian=20G=C3=B6ttsche?= <cgzones@googlemail.com>
+Date: Mon, 3 May 2021 17:11:09 +0200
+Subject: [PATCH] libselinux: store_stem(): do not free possible non-heap
+ object
+MIME-Version: 1.0
+Content-Type: text/plain; charset=UTF-8
+Content-Transfer-Encoding: 8bit
+
+GCC 11 complains:
+
+In file included from label_file.c:24:
+In function ‘store_stem’,
+    inlined from ‘load_mmap’ at label_file.c:277:12,
+    inlined from ‘process_file’ at label_file.c:551:5:
+label_file.h:289:25: error: ‘free’ called on pointer ‘*mmap_area.next_addr’ with nonzero offset 4 [-Werror=free-nonheap-object]
+  289 |                         free(buf);
+      |                         ^~~~~~~~~
+
+Free the pointer on failure at the caller instead of inside `store_stem()`.
+
+Signed-off-by: Christian Göttsche <cgzones@googlemail.com>
+---
+ libselinux/src/label_file.h | 8 ++++++--
+ 1 file changed, 6 insertions(+), 2 deletions(-)
+
+diff --git a/libselinux/src/label_file.h b/libselinux/src/label_file.h
+index baed3341..9f633701 100644
+--- a/libselinux/src/label_file.h
++++ b/libselinux/src/label_file.h
+@@ -286,7 +286,6 @@ static inline int store_stem(struct saved_data *data, char *buf, int stem_len)
+ 		tmp_arr = realloc(data->stem_arr,
+ 				  sizeof(*tmp_arr) * alloc_stems);
+ 		if (!tmp_arr) {
+-			free(buf);
+ 			return -1;
+ 		}
+ 		data->alloc_stems = alloc_stems;
+@@ -308,6 +307,7 @@ static inline int find_stem_from_spec(struct saved_data *data, const char *buf)
+ 	int stem_len = get_stem_from_spec(buf);
+ 	int stemid;
+ 	char *stem;
++	int r;
+ 
+ 	if (!stem_len)
+ 		return -1;
+@@ -321,7 +321,11 @@ static inline int find_stem_from_spec(struct saved_data *data, const char *buf)
+ 	if (!stem)
+ 		return -1;
+ 
+-	return store_stem(data, stem, stem_len);
++	r = store_stem(data, stem, stem_len);
++	if (r < 0)
++		free(stem);
++
++	return r;
+ }
+ 
+ /* This will always check for buffer over-runs and either read the next entry

